File: python/lesson04/solution/hello.py
import requests
import sys
import time
from lib.tracing import init_tracer
from opentracing_instrumentation.request_context import get_current_span, span_in_context
from opentracing.ext import tags
from opentracing.propagation import Format


def say_hello(hello_to, greeting):
    with tracer.start_span('say-hello') as span:
        span.set_tag('hello-to', hello_to)
        span.set_baggage_item('greeting', greeting)
        with span_in_context(span):
            hello_str = format_string(hello_to)
            print_hello(hello_str)

def format_string(hello_to):
    with tracer.start_span('format', child_of=get_current_span()) as span:
        with span_in_context(span):
            hello_str = http_get(8081, 'format', 'helloTo', hello_to)
            span.log_kv({'event': 'string-format', 'value': hello_str})
            return hello_str

def print_hello(hello_str):
    with tracer.start_span('println', child_of=get_current_span()) as span:
        with span_in_context(span):
            http_get(8082, 'publish', 'helloStr', hello_str)
            span.log_kv({'event': 'println'})

def http_get(port, path, param, value):
    url = 'http://localhost:%s/%s' % (port, path)

    span = get_current_span()
    span.set_tag(tags.HTTP_METHOD, 'GET')
    span.set_tag(tags.HTTP_URL, url)
    span.set_tag(tags.SPAN_KIND, tags.SPAN_KIND_RPC_CLIENT)
    headers = {}
    tracer.inject(span, Format.HTTP_HEADERS, headers)

    r = requests.get(url, params={param: value}, headers=headers)
    assert r.status_code == 200
    return r.text
        

# main
assert len(sys.argv) == 3

tracer = init_tracer('hello-world')

hello_to = sys.argv[1]
greeting = sys.argv[2]
say_hello(hello_to, greeting)

# yield to IOLoop to flush the spans
time.sleep(2)
tracer.close()

File: nodejs/lesson04/solution/hello.js
'use strict';

const assert = require('assert');
const initTracer = require('../../lib/tracing').initTracer;
const request = require('request-promise');
const { Tags, FORMAT_HTTP_HEADERS } = require('opentracing');

function sayHello(helloTo, greeting) {

    const span = tracer.startSpan('say-hello');
    span.setTag('hello-to', helloTo);

    span.setBaggageItem('greeting', greeting)

    format_string(helloTo, span)
        .then( data => {
            return print_hello(data, span);
        })
        .then( data => {
            span.setTag(Tags.HTTP_STATUS_CODE, 200)
            span.finish();
        })
        .catch( err => {
            span.setTag(Tags.ERROR, true) 
            span.setTag(Tags.HTTP_STATUS_CODE, err.statusCode || 500);
            span.finish();
            throw err;
        });

}

function format_string(input, root_span) {
    const url = `http://localhost:8081/format?helloTo=${input}`;
    const fn = 'format';

    const span = tracer.startSpan(fn, {childOf: root_span.context()});
    span.log({
        'event': 'format-string',
        'value': input
    });

    return http_get(url, span); 
}  

function print_hello(input, root_span) {
    const url = `http://localhost:8082/publish?helloStr=${input}`;
    const fn = 'publish';

    const span = tracer.startSpan(fn, {childOf: root_span.context()});
    span.log({
        'event': 'print-string',
        'value': input
    });
    return http_get(url, span);
}

function http_get(url, span) {
    const method = 'GET';
    const headers = {};
    
    span.setTag(Tags.HTTP_URL, url);
    span.setTag(Tags.HTTP_METHOD, method);
    span.setTag(Tags.SPAN_KIND, Tags.SPAN_KIND_RPC_CLIENT);
    // Send span context via request headers (parent id etc.)
    tracer.inject(span, FORMAT_HTTP_HEADERS, headers);

    return request({url, method, headers})
            .then( data => {
                span.finish();
                return data;
            }, e => {
                span.setTag(Tags.ERROR, true)
                span.log({
                    'event': 'error',
                    'error.object': e
                });
                span.finish();
                throw e;
            });

}

assert.ok(process.argv.length == 4, 'expecting two argument');

const [, , helloTo, greeting] = process.argv;

const tracer = initTracer('hello-world');

sayHello(helloTo, greeting);

setTimeout( e => {tracer.close();}, 12000);
